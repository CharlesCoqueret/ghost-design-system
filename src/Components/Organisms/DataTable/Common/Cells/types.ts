import {
  IColumnType,
  IExtraEditableDataTableProps,
  IExtraLineEditableInPlaceDataTableProps,
  IExtraStaticDataTableProps,
} from '../types';

export interface ICellProps<T, U extends IColumnType<T>> {
  column: U;
  dataTestId?: string;
  editing?: boolean;
  extra?: IExtraStaticDataTableProps<T> | IExtraLineEditableInPlaceDataTableProps<T> | IExtraEditableDataTableProps<T>;
  forcedValue?: T[keyof T] | number | string | Date | null;
  onChange?: (newValue: T[keyof T]) => void;
  row?: T;
  rowIndex: number;
}
