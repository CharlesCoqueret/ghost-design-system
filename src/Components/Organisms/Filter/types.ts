import { ThousandsGroupStyle } from '../../Atoms/AmountInput';
import { DateFormat, WeekDayEnum } from '../../Atoms/DatePickerInput';
import { IOption } from '../../Atoms/SelectInput';

export enum FilterTypeEnum {
  CHECKBOX = 'checkbox',
  DATE = 'date',
  DYNAMICSEARCH = 'dynamicsearch',
  MULTISELECT = 'multiselect',
  NUMBER = 'number',
  COLUMN = 'column',
  SELECT = 'select',
  TEXT = 'text',
  TITLE = 'title',
}

export type IFilterLayoutAndFieldsProps<T> = IFilerColProps<T> | IFilerTitleProps | IFilterFieldsProps<T>;

export type IFilterFieldsProps<T> =
  | IFilterCheckboxProps<T>
  | IFilterDateProps<T>
  | IFilterDynamicSearchProps<T>
  | IFilterMultiSelectProps<T>
  | IFilterNumberProps<T>
  | IFilterSelectProps<T>
  | IFilterTextProps<T>;

export interface IFilterBaseProps<T> {
  dataIndex: keyof T;
  dataTestId?: string;
  fieldClassName?: string;
  fieldSize?: number;
  helperText?: string;
  inputClassName?: string;
  label?: string;
}

export interface IFilterCheckboxProps<T> extends IFilterBaseProps<T> {
  filterType: FilterTypeEnum.CHECKBOX;
}

export interface IFilterDateProps<T> extends IFilterBaseProps<T> {
  calendarStartDay?: WeekDayEnum;
  dateFormat?: DateFormat;
  filterType: FilterTypeEnum.DATE;
  isClearable?: boolean;
  locale?: string;
  name?: string;
  placeholder?: string;
}

export interface IFilterDynamicSearchProps<T> extends IFilterBaseProps<T> {
  colors?: {
    controlErrorColor: string; // colors.error,
    controlFocusColor: string; // colors.primary,
    fontColor: string; // 'rgb(0, 0, 0)',
    optionFocusColor: string; // colors.chalk,
    optionSelectedColor: string; // colors.primary,
  };
  filterType: FilterTypeEnum.DYNAMICSEARCH;
  isClearable?: boolean;
  name?: string;
  noOptionsMessage: (obj: { inputValue: string }) => string;
  placeholder?: string;
  resolveValue: (value: string | number) => Promise<IOption | undefined>;
  searchOptions: (searchTerm: string) => Promise<Array<IOption> | undefined>;
}

export interface IFilterMultiSelectProps<T> extends IFilterBaseProps<T> {
  colors?: {
    controlErrorColor: string; // colors.error,
    controlFocusColor: string; // colors.primary,
    fontColor: string; // 'rgb(0, 0, 0)',
    optionFocusColor: string; // colors.chalk,
    optionSelectedColor: string; // colors.primary,
  };
  filterType: FilterTypeEnum.MULTISELECT;
  isClearable?: boolean;
  name?: string;
  numberOfItemLabel: string;
  numberOfItemsLabel: string;
  options: Array<IOption>;
  placeholder?: string;
}

export interface IFilterNumberProps<T> extends IFilterBaseProps<T> {
  allowNegative?: boolean;
  decimalScale?: number;
  decimalSeparator?: string;
  filterType: FilterTypeEnum.NUMBER;
  maxValue?: number;
  minValue?: number;
  name?: string;
  placeholder?: string;
  prefix?: string | undefined;
  suffix?: string | undefined;
  thousandSeparator?: string;
  thousandsGroupStyle?: ThousandsGroupStyle;
}

export interface IFilerColProps<T> {
  filterType: FilterTypeEnum.COLUMN;
  fields: Array<IFilterLayoutAndFieldsProps<T>>;
}

export interface IFilterSelectProps<T> extends IFilterBaseProps<T> {
  colors?: {
    controlErrorColor: string; // colors.error,
    controlFocusColor: string; // colors.primary,
    fontColor: string; // 'rgb(0, 0, 0)',
    optionFocusColor: string; // colors.chalk,
    optionSelectedColor: string; // colors.primary,
  };
  filterType: FilterTypeEnum.SELECT;
  isClearable?: boolean;
  name?: string;
  options: Array<IOption>;
  placeholder?: string;
}

export interface IFilterTextProps<T> extends IFilterBaseProps<T> {
  filterType: FilterTypeEnum.TEXT;
  maxLength?: number;
  minLength?: number;
  name?: string;
  placeholder?: string;
}

export interface IFilerTitleProps {
  filterType: FilterTypeEnum.TITLE;
  label: string;
}
