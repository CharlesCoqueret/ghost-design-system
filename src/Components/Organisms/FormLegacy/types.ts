import { ReactElement } from 'react';
import * as yup from 'yup';

import { IOption } from '../../Atoms/SelectInput';
import { IAmountFieldProps } from '../../Molecules/AmountField/AmountField';
import { IDatePickerFieldProps } from '../../Molecules/DatePickerField/DatePickerField';
import { IFileFieldProps } from '../../Molecules/FileField/FileField';
import { IDynamicSearchFieldProps } from '../../Molecules/DynamicSearchField/DynamicSearchField';
import { IMultiSelectFieldProps } from '../../Molecules/MultiSelectField/MultiSelectField';
import { IPercentageFieldProps } from '../../Molecules/PercentageField/PercentageField';
import { IRichTextFieldProps } from '../../Molecules/RichTextField/RichTextField';
import { ISelectFieldProps } from '../../Molecules/SelectField/SelectField';
import { ILineEditableDataTableProps } from '../DataTable/LineEditableDataTable/LineEditableDataTable';
import { IEditableDataTableProps } from '../DataTable/EditableDataTable/EditableDataTable';
import { ITextFieldProps } from '../../Molecules/TextField/TextField';
import { ITextAreaFieldProps } from '../../Molecules/TextAreaField/TextAreaField';
import { IYearPickerFieldProps } from '../../Molecules/YearPickerField/YearPickerField';
import { IFormProps } from './FormLegacy';

export interface IFormLegacySubmitReturnedType<T> {
  data: T;
  valid: boolean;
}

export interface IUseFormLegacyReturnedType<T extends yup.AnyObject> {
  formElement: ReactElement;
  formProps: IFormProps<T>;
  getData: () => T;
  isModified: () => boolean;
  rehydrate: (data: T) => void;
  reset: () => void;
  submit: () => IFormLegacySubmitReturnedType<T>;
}

export enum FieldLegacyTypeEnum {
  /** Amount field */
  AMOUNT = 'amount',
  /** Checkbox field */
  CHECKBOX = 'checkbox',
  /**
   * Custom field
   * @deprecated will be removed in version 2.1, please use useForm for custom fields
   */
  CUSTOM = 'custom',
  /** Date field */
  DATE = 'date',
  /** Description field */
  DESCRIPTION = 'description',
  /** Dynamic search field */
  DYNAMICSEARCH = 'dynamicsearch',
  /** File field */
  FILE = 'file',
  /** Multi select field */
  MULTISELECT = 'multiselect',
  /** Number field */
  NUMBER = 'number',
  /** Percentage field */
  PERCENTAGE = 'percentage',
  /** Rich text field */
  RICHTEXT = 'richtext',
  /** Section field */
  SECTION = 'section',
  /** Select field */
  SELECT = 'select',
  /** Switch field */
  SWITCH = 'switch',
  /**
   * Line editable table
   * @deprecated will be removed in version 2.1, please use useForm for line editable table
   */
  LINE_EDITABLE_TABLE = 'lineeditabletable',
  /**
   * Editable table
   * @deprecated will be removed in version 2.1, please use useForm for line editable table
   */
  EDITABLE_TABLE = 'editabletable',
  /** Text field */
  TEXT = 'text',
  /** Text area field */
  TEXTAREA = 'textarea',
  /** Year field */
  YEAR = 'year',
}

export type IFieldAndLayoutLegacyProps<T> = IFieldLegacyProps<T> | ILayoutLegacyProps<T>;

export type ILayoutLegacyProps<T> = IFieldSectionLegacyProps<T> | IFieldDescriptionLegacyProps<T>;

export type IFieldLegacyProps<T> =
  | IFieldAmountProps<T>
  | IFieldCheckboxProps<T>
  | IFieldCustomProps<T>
  | IFieldDateProps<T>
  | IFieldDynamicSearchProps<T>
  | IFieldFileProps<T>
  | IFieldMultiSelectProps<T>
  | IFieldNumberProps<T>
  | IFieldPercentageProps<T>
  | IFieldRichtextProps<T>
  | IFieldSelectProps<T>
  | IFieldSwitchProps<T>
  // TODO investigate type resolution
  // Using any to avoid circular type definition for now, until there is a way to get the type of an item of T[keyof T]
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  | IFieldLineEditableTableProps<T, any>
  // TODO investigate type resolution
  // Using any to avoid circular type definition for now, until there is a way to get the type of an item of T[keyof T]
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  | IFieldEditableTableProps<T, any>
  | IFieldTextAreaProps<T>
  | IFieldTextProps<T>
  | IFieldYearProps<T>;

export interface IVisibilityProps<T> {
  /** When set to true or when called, returning true, it hides the field or description or section (and children fields) (optional, default: false) */
  hidden?: boolean | ((data: T) => boolean);
}

export interface IFieldBaseProps<T> extends Partial<IVisibilityProps<T>> {
  dataIndex: keyof T;
  disabled?: boolean;
  errorMessage?: string;
  fieldClassName?: string;
  helperText?: string;
  highlighted?: boolean;
  inputClassName?: string;
  label: string;
  mandatory?: boolean;
  name?: string;
  readOnly?: boolean;
}

export interface IFieldAmountProps<T>
  extends IFieldBaseProps<T>,
    Pick<
      IAmountFieldProps,
      | 'allowNegative'
      | 'decimalScale'
      | 'decimalSeparator'
      | 'maxValue'
      | 'minValue'
      | 'placeholder'
      | 'prefix'
      | 'suffix'
      | 'thousandSeparator'
      | 'thousandsGroupStyle'
    > {
  fieldType: FieldLegacyTypeEnum.AMOUNT;
}

export interface IFieldCheckboxProps<T> extends IFieldBaseProps<T> {
  fieldType: FieldLegacyTypeEnum.CHECKBOX;
}

export interface IFieldCustomProps<T, U = unknown> extends IFieldBaseProps<T> {
  fieldType: FieldLegacyTypeEnum.CUSTOM;
  isEqual?: (previousValue: T[keyof T], currentValue: T[keyof T]) => boolean;
  data?: U;
  customField: <
    U extends {
      highlighted?: boolean;
      onChange?: (value: T[keyof T]) => void;
      readOnly?: boolean;
      input?: T[keyof T];
    },
  >(
    props: U,
  ) => ReactElement;
}

export interface IFieldDateProps<T>
  extends IFieldBaseProps<T>,
    Pick<IDatePickerFieldProps, 'calendarStartDay' | 'dateFormat' | 'isClearable' | 'locale' | 'placeholder'> {
  fieldType: FieldLegacyTypeEnum.DATE;
}

export interface IFieldDescriptionLegacyProps<T> extends Partial<IVisibilityProps<T>> {
  dataIndex?: keyof T;
  description: ReactElement | (<U extends { value: T[keyof T] }>(props: U) => ReactElement);
  fieldType: FieldLegacyTypeEnum.DESCRIPTION;
}

export interface IFieldFileProps<T>
  extends IFieldBaseProps<T>,
    Pick<
      IFileFieldProps,
      | 'acceptTypes'
      | 'additionalInfo'
      | 'maxFiles'
      | 'maxFileSize'
      | 'maxFolderDepth'
      | 'onDelete'
      | 'onDownload'
      | 'onFailure'
      | 'onSuccess'
      | 'requestHeaders'
      | 'requestMethod'
      | 'requestUrl'
      | 'requestWithCredentials'
      | 'showFileSize'
      | 'showProgressBar'
      | 'uploadMessage'
      | 'localization'
    > {
  fieldType: FieldLegacyTypeEnum.FILE;
}

export interface IFieldDynamicSearchProps<T>
  extends IFieldBaseProps<T>,
    Pick<
      IDynamicSearchFieldProps,
      'isClearable' | 'noOptionsMessage' | 'placeholder' | 'resolveValue' | 'searchOptions'
    > {
  fieldType: FieldLegacyTypeEnum.DYNAMICSEARCH;
}

export interface IFieldMultiSelectProps<T>
  extends IFieldBaseProps<T>,
    Pick<IMultiSelectFieldProps, 'isClearable' | 'numberOfItemLabel' | 'numberOfItemsLabel' | 'placeholder'> {
  // When the value is not present in the options, should the value be erased (optional, default: false)
  eraseValueWhenNotInOptions?: boolean;
  fieldType: FieldLegacyTypeEnum.MULTISELECT;
  options: Array<IOption> | ((data: T) => Array<IOption>);
}

export interface IFieldNumberProps<T>
  extends IFieldBaseProps<T>,
    Pick<
      IAmountFieldProps,
      | 'allowNegative'
      | 'decimalScale'
      | 'decimalSeparator'
      | 'maxValue'
      | 'minValue'
      | 'placeholder'
      | 'prefix'
      | 'suffix'
      | 'thousandSeparator'
      | 'thousandsGroupStyle'
    > {
  fieldType: FieldLegacyTypeEnum.NUMBER;
}

export interface IFieldPercentageProps<T>
  extends IFieldBaseProps<T>,
    Pick<
      IPercentageFieldProps,
      | 'allowNegative'
      | 'decimalScale'
      | 'decimalSeparator'
      | 'maxValue'
      | 'minValue'
      | 'placeholder'
      | 'thousandSeparator'
      | 'thousandsGroupStyle'
    > {
  fieldType: FieldLegacyTypeEnum.PERCENTAGE;
}

export interface IFieldRichtextProps<T>
  extends IFieldBaseProps<T>,
    Pick<IRichTextFieldProps, 'convertImagesToBase64' | 'enableImage' | 'enableLink' | 'locale' | 'maxLength'> {
  fieldType: FieldLegacyTypeEnum.RICHTEXT;
}

export interface IFieldSectionLegacyProps<T> extends Partial<IVisibilityProps<T>> {
  collapsible?: boolean;
  fieldType: FieldLegacyTypeEnum.SECTION;
  fields: Array<IFieldLegacyProps<T>>;
  label: string;
  level?: 1 | 2 | 3;
  openInitially?: boolean;
  separator?: boolean;
}

export interface IFieldSelectProps<T>
  extends IFieldBaseProps<T>,
    Pick<ISelectFieldProps, 'isClearable' | 'placeholder'> {
  // When the value is not present in the options, should the value be erased (optional, default: false)
  eraseValueWhenNotInOptions?: boolean;
  fieldType: FieldLegacyTypeEnum.SELECT;
  options: Array<IOption> | ((data: T) => Array<IOption>);
}

export interface IFieldSwitchProps<T> extends IFieldBaseProps<T> {
  fieldType: FieldLegacyTypeEnum.SWITCH;
}

export interface IFieldEditableTableProps<T, U extends yup.AnyObject>
  extends IFieldBaseProps<T>,
    Pick<IEditableDataTableProps<U>, 'columns' | 'extra' | 'loading' | 'onSortChange'> {
  fieldType: FieldLegacyTypeEnum.EDITABLE_TABLE;
}
export interface IFieldLineEditableTableProps<T, U extends yup.AnyObject>
  extends IFieldBaseProps<T>,
    Pick<ILineEditableDataTableProps<U>, 'columns' | 'extra' | 'loading' | 'onSortChange'> {
  fieldType: FieldLegacyTypeEnum.LINE_EDITABLE_TABLE;
}

export interface IFieldTextProps<T>
  extends IFieldBaseProps<T>,
    Pick<ITextFieldProps, 'maxLength' | 'minLength' | 'placeholder'> {
  fieldType: FieldLegacyTypeEnum.TEXT;
}

export interface IFieldTextAreaProps<T>
  extends IFieldBaseProps<T>,
    Pick<ITextAreaFieldProps, 'maxLength' | 'minLength' | 'placeholder'> {
  fieldType: FieldLegacyTypeEnum.TEXTAREA;
}

export interface IFieldYearProps<T> extends IFieldBaseProps<T>, Pick<IYearPickerFieldProps, 'placeholder'> {
  fieldType: FieldLegacyTypeEnum.YEAR;
}
