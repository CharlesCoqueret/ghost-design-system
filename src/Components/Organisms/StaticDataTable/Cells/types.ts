import { IColumnType, IExtraLineEditableDataTableProps, IExtraStaticDataTableProps } from '../types';

export interface ICellProps<T, U extends IColumnType<T>> {
  column: U;
  row?: T;
  rowIndex: number;
  extra?: IExtraStaticDataTableProps<T> | IExtraLineEditableDataTableProps<T>;
  forcedValue?: T[keyof T] | number | string | Date | null;
  onChange?: (newValue?: T[keyof T]) => void;
}
