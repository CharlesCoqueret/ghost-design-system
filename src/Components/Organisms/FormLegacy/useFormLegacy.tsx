import React, { useState } from 'react';
import cloneDeep from 'lodash/cloneDeep';
import * as yup from 'yup';

import FormLegacy from './FormLegacy';
import { IFieldAndLayoutLegacyProps, IFormLegacySubmitReturnedType, IUseFormLegacyReturnedType } from './types';
import { yupResolver } from './yupResolver';
import { useRunAfterUpdate } from '../../../hooks';

export interface IUseFormProps<T extends yup.AnyObject> {
  enableOldData?: boolean;
  enableSideBySide?: boolean;
  fields: Array<IFieldAndLayoutLegacyProps<T>>;
  initialData: T;
  onChangeNotification?: () => void;
  previousData?: T;
  usePortal?: boolean;
  validationSchema: yup.ObjectSchema<T>;
}

const useFormLegacy = <T extends yup.AnyObject>(props: IUseFormProps<T>): IUseFormLegacyReturnedType<T> => {
  const {
    enableOldData,
    enableSideBySide,
    fields,
    initialData,
    onChangeNotification,
    previousData,
    usePortal,
    validationSchema,
  } = props;

  const [currentData, setCurrentData] = useState<T>(cloneDeep(initialData));
  const [isModified, setIsModified] = useState<boolean>(false);
  const [validationError, setValidationError] = useState<Partial<Record<keyof T, string>>>();
  const runAfterUpdate = useRunAfterUpdate();

  const rehydrate = (newData: T): void => {
    setCurrentData(newData);
    setIsModified(false);
    setValidationError(undefined);
  };

  const handleDataChange = (dataIndex: keyof T, newValue: T[keyof T]): void => {
    setCurrentData((prev) => {
      prev[dataIndex] = newValue;
      return { ...prev };
    });

    if (onChangeNotification) {
      onChangeNotification();
    }

    setValidationError((prev) => {
      if (prev === undefined) return prev;
      delete prev[dataIndex];
      return { ...prev };
    });
    setIsModified(true);
  };

  const getData = () => {
    return cloneDeep(currentData);
  };

  const reset = () => {
    setCurrentData(cloneDeep(initialData));

    if (onChangeNotification) {
      onChangeNotification();
    }

    setValidationError(undefined);
    setIsModified(false);
  };

  const scrollToError = () => {
    const error = document.body.getElementsByClassName('field-error').item(0);
    if (!error) return;
    error.scrollIntoView({ behavior: 'smooth', block: 'center' });
  };

  const submit = (): IFormLegacySubmitReturnedType<T> => {
    let isValid = true;
    if (validationSchema) {
      const errors = yupResolver(validationSchema, { abortEarly: false }, currentData);
      isValid = errors === undefined;
      // Logging validation errors in case the developer has checked it.
      if (!isValid) console.error(errors);
      setValidationError(errors);
    }

    if (!isValid) {
      runAfterUpdate(scrollToError);
    }

    return { data: cloneDeep(currentData), valid: isValid };
  };

  return {
    formElement: (
      <FormLegacy
        enableOldData={enableOldData}
        enableSideBySide={enableSideBySide}
        fields={fields}
        handleDataChange={handleDataChange}
        initialData={currentData}
        validationError={validationError}
        previousData={previousData}
        validationSchema={validationSchema}
        usePortal={usePortal}
      />
    ),
    formProps: {
      enableOldData,
      enableSideBySide,
      fields,
      handleDataChange,
      initialData: currentData,
      validationError,
      previousData,
      validationSchema,
      usePortal,
    },
    getData,
    isModified: () => isModified,
    rehydrate,
    reset,
    submit,
  };
};

useFormLegacy.defaultProps = {
  enableOldData: undefined,
  enableSideBySide: undefined,
  onChangeNotification: undefined,
  previousData: undefined,
  validationSchema: undefined,
  usePortal: true,
};

export default useFormLegacy;
