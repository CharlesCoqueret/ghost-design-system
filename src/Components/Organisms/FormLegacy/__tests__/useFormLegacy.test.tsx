import React from 'react';
import { act, render, renderHook } from '@testing-library/react';

import * as yup from 'yup';
import { FieldLegacyTypeEnum } from '../types';

let handleDataChangeForm: ((dataIndex: string, newValue: number | undefined) => void) | undefined;

jest.mock('../FormLegacy', () => {
  return {
    __esModule: true,
    default: (props: { handleDataChange: typeof handleDataChangeForm; validationError?: Record<string, unknown> }) => {
      if (props.handleDataChange) handleDataChangeForm = props.handleDataChange;
      return (
        <>
          <div>Mocked Form props: {JSON.stringify(props)}</div>
          {props.validationError && <div className='field-error' />}
        </>
      );
    },
  };
});

import useFormLegacy from '../useFormLegacy';

describe('useFormLegacy hook', () => {
  it('provides the right elements', () => {
    //scrollIntoView is not implemented in jsdom
    window.HTMLElement.prototype.scrollIntoView = jest.fn();

    console.error = jest.fn();

    const { result } = renderHook(() =>
      useFormLegacy({
        initialData: { number: 50 },
        fields: [{ fieldType: FieldLegacyTypeEnum.NUMBER, dataIndex: 'number', label: 'Number' }],
        previousData: { number: 0 },
        validationSchema: yup.object({ number: yup.number().required('Value for number is required') }),
      }),
    );

    const { container, rerender } = render(result.current.formElement);
    expect(container).toMatchSnapshot();
    expect(result.current.getData()).toEqual({ number: 50 });
    expect(result.current.isModified()).toEqual(false);

    act(() => {
      if (handleDataChangeForm) {
        handleDataChangeForm('number', undefined);
      }
    });
    rerender(result.current.formElement);
    expect(result.current.getData()).toEqual({ number: undefined });
    expect(result.current.isModified()).toEqual(true);
    expect(container).toMatchSnapshot();

    act(() => {
      expect(result.current.submit()).toEqual({ valid: false, data: { number: undefined } });
    });
    expect(console.error).toBeCalledTimes(1);
    expect(console.error).toBeCalledWith({ number: 'Value for number is required' });

    rerender(result.current.formElement);
    act(() => {
      // Running once more as the rerender is managed manually.
      result.current.submit();
    });

    // eslint-disable-next-line @typescript-eslint/unbound-method
    expect(window.HTMLElement.prototype.scrollIntoView).toBeCalledTimes(1);
    expect(container).toMatchSnapshot();
    expect(console.error).toBeCalledTimes(2);

    act(() => {
      if (handleDataChangeForm) {
        handleDataChangeForm('number', 100);
      }
    });
    rerender(result.current.formElement);
    expect(container).toMatchSnapshot();
    expect(result.current.getData()).toEqual({ number: 100 });
    expect(result.current.isModified()).toEqual(true);

    act(() => {
      expect(result.current.submit()).toEqual({ valid: true, data: { number: 100 } });
    });
    rerender(result.current.formElement);
    expect(container).toMatchSnapshot();
    expect(result.current.getData()).toEqual({ number: 100 });
    expect(result.current.isModified()).toEqual(true);

    act(() => {
      result.current.reset();
    });
    rerender(result.current.formElement);
    expect(container).toMatchSnapshot();
    expect(result.current.getData()).toEqual({ number: 50 });
    expect(result.current.isModified()).toEqual(false);
  });
});
