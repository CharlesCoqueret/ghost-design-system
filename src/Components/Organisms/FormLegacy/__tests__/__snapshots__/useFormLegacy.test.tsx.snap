// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFormLegacy hook provides the right elements 1`] = `
<div>
  <div>
    Mocked Form props: 
    {"fields":[{"fieldType":"number","dataIndex":"number","label":"Number"}],"initialData":{"number":50},"previousData":{"number":0},"validationSchema":{"type":"object","deps":[],"tests":[],"transforms":[],"conditions":[],"internalTests":{},"_whitelist":{},"_blacklist":{},"exclusiveTests":{},"spec":{"strip":false,"strict":false,"abortEarly":true,"recursive":true,"nullable":false,"optional":true,"coerce":true},"fields":{"number":{"type":"number","_whitelist":{},"_blacklist":{},"internalTests":{},"exclusiveTests":{},"deps":[],"conditions":[],"tests":[],"transforms":[null],"spec":{"strip":false,"strict":false,"abortEarly":true,"recursive":true,"nullable":false,"optional":false,"coerce":true}}},"_nodes":["number"],"_excludedEdges":[]}}
  </div>
</div>
`;

exports[`useFormLegacy hook provides the right elements 2`] = `
<div>
  <div>
    Mocked Form props: 
    {"fields":[{"fieldType":"number","dataIndex":"number","label":"Number"}],"initialData":{},"previousData":{"number":0},"validationSchema":{"type":"object","deps":[],"tests":[],"transforms":[],"conditions":[],"internalTests":{},"_whitelist":{},"_blacklist":{},"exclusiveTests":{},"spec":{"strip":false,"strict":false,"abortEarly":true,"recursive":true,"nullable":false,"optional":true,"coerce":true},"fields":{"number":{"type":"number","_whitelist":{},"_blacklist":{},"internalTests":{},"exclusiveTests":{},"deps":[],"conditions":[],"tests":[],"transforms":[null],"spec":{"strip":false,"strict":false,"abortEarly":true,"recursive":true,"nullable":false,"optional":false,"coerce":true}}},"_nodes":["number"],"_excludedEdges":[]}}
  </div>
</div>
`;

exports[`useFormLegacy hook provides the right elements 3`] = `
<div>
  <div>
    Mocked Form props: 
    {"fields":[{"fieldType":"number","dataIndex":"number","label":"Number"}],"initialData":{},"validationError":{"number":"Value for number is required"},"previousData":{"number":0},"validationSchema":{"type":"object","deps":[],"tests":[],"transforms":[],"conditions":[],"internalTests":{},"_whitelist":{},"_blacklist":{},"exclusiveTests":{},"spec":{"strip":false,"strict":false,"abortEarly":true,"recursive":true,"nullable":false,"optional":true,"coerce":true},"fields":{"number":{"type":"number","_whitelist":{},"_blacklist":{},"internalTests":{},"exclusiveTests":{},"deps":[],"conditions":[],"tests":[],"transforms":[null],"spec":{"strip":false,"strict":false,"abortEarly":true,"recursive":true,"nullable":false,"optional":false,"coerce":true}}},"_nodes":["number"],"_excludedEdges":[]}}
  </div>
  <div
    class="field-error"
  />
</div>
`;

exports[`useFormLegacy hook provides the right elements 4`] = `
<div>
  <div>
    Mocked Form props: 
    {"fields":[{"fieldType":"number","dataIndex":"number","label":"Number"}],"initialData":{"number":100},"validationError":{},"previousData":{"number":0},"validationSchema":{"type":"object","deps":[],"tests":[],"transforms":[],"conditions":[],"internalTests":{},"_whitelist":{},"_blacklist":{},"exclusiveTests":{},"spec":{"strip":false,"strict":false,"abortEarly":true,"recursive":true,"nullable":false,"optional":true,"coerce":true},"fields":{"number":{"type":"number","_whitelist":{},"_blacklist":{},"internalTests":{},"exclusiveTests":{},"deps":[],"conditions":[],"tests":[],"transforms":[null],"spec":{"strip":false,"strict":false,"abortEarly":true,"recursive":true,"nullable":false,"optional":false,"coerce":true}}},"_nodes":["number"],"_excludedEdges":[]}}
  </div>
  <div
    class="field-error"
  />
</div>
`;

exports[`useFormLegacy hook provides the right elements 5`] = `
<div>
  <div>
    Mocked Form props: 
    {"fields":[{"fieldType":"number","dataIndex":"number","label":"Number"}],"initialData":{"number":100},"previousData":{"number":0},"validationSchema":{"type":"object","deps":[],"tests":[],"transforms":[],"conditions":[],"internalTests":{},"_whitelist":{},"_blacklist":{},"exclusiveTests":{},"spec":{"strip":false,"strict":false,"abortEarly":true,"recursive":true,"nullable":false,"optional":true,"coerce":true},"fields":{"number":{"type":"number","_whitelist":{},"_blacklist":{},"internalTests":{},"exclusiveTests":{},"deps":[],"conditions":[],"tests":[],"transforms":[null],"spec":{"strip":false,"strict":false,"abortEarly":true,"recursive":true,"nullable":false,"optional":false,"coerce":true}}},"_nodes":["number"],"_excludedEdges":[]}}
  </div>
</div>
`;

exports[`useFormLegacy hook provides the right elements 6`] = `
<div>
  <div>
    Mocked Form props: 
    {"fields":[{"fieldType":"number","dataIndex":"number","label":"Number"}],"initialData":{"number":50},"previousData":{"number":0},"validationSchema":{"type":"object","deps":[],"tests":[],"transforms":[],"conditions":[],"internalTests":{},"_whitelist":{},"_blacklist":{},"exclusiveTests":{},"spec":{"strip":false,"strict":false,"abortEarly":true,"recursive":true,"nullable":false,"optional":true,"coerce":true},"fields":{"number":{"type":"number","_whitelist":{},"_blacklist":{},"internalTests":{},"exclusiveTests":{},"deps":[],"conditions":[],"tests":[],"transforms":[null],"spec":{"strip":false,"strict":false,"abortEarly":true,"recursive":true,"nullable":false,"optional":false,"coerce":true}}},"_nodes":["number"],"_excludedEdges":[]}}
  </div>
</div>
`;
